sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisReader.java

```diff
@@ -84,11 +85,17 @@ public boolean advance() throws IOException {
       for (int i = 0; i < shardIterators.size(); ++i) {
         currentRecord = shardIterators.getCurrent().next();
         if (currentRecord.isPresent()) {
+          Instant approximateArrivalTimestamp = currentRecord.get()
+              .getApproximateArrivalTimestamp();
+          if (approximateArrivalTimestamp.isAfter(watermark)) {
```

Comment:
So we declare that watermark is the latest arrival timestamp seen so far - this seems like a very aggressive strategy for advancing the watermark, because it means that if records are even slightly out of order, they will be declared late data. PubsubIO uses a less aggressive strategy: https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubUnboundedSource.java#L88 - would it be a good idea to use something similar here? (if you believe the current PR is already a vast improvement over the state of the code before this PR, then I'm fine getting it in as-is, but it's something worth thinking about).

Reply:
It sounds like a really good idea to implement it in a similar way. I'll prepare the changes shortly.